File: NotificationAPI.API/Controllers/UsersController.cs
using AutoMapper;
using Microsoft.AspNetCore.Mvc;
using NotificationAPI.Application.DTOs.Requests;
using NotificationAPI.Application.DTOs.Responses;
using NotificationAPI.Application.UseCases.Users;
using NotificationAPI.Domain.Entities;
using NotificationAPI.Domain.Repositories;

namespace NotificationAPI.API.Controllers;

[ApiController]
[Route("api/[controller]")]
public class UsersController : ControllerBase
{
    private readonly IUnitOfWork _unitOfWork;
    private readonly IMapper _mapper;
    private readonly GetUserByIdUseCase _getUserByIdUseCase;

    public UsersController(IUnitOfWork unitOfWork, IMapper mapper, GetUserByIdUseCase getUserByIdUseCase)
    {
        _unitOfWork = unitOfWork;
        _mapper = mapper;
        _getUserByIdUseCase = getUserByIdUseCase;
    }

    [HttpPost]
    [ProducesResponseType(typeof(UserResponse), StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> CreateUser([FromBody] CreateUserRequest request)
    {
        try
        {
            var user = _mapper.Map<User>(request);
            await _unitOfWork.Users.AddAsync(user);
            await _unitOfWork.SaveChangesAsync();

            var response = _mapper.Map<UserResponse>(user);
            return CreatedAtAction(nameof(GetUserById), new { id = response.Id }, response);
        }
        catch (Exception)
        {
            return StatusCode(StatusCodes.Status500InternalServerError, new { error = "An unexpected error occurred" });
        }
    }

    [HttpGet("{id:guid}")]
    [ProducesResponseType(typeof(UserResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetUserById(Guid id)
    {
        try
        {
            var response = await _getUserByIdUseCase.ExecuteAsync(id);
            return Ok(response);
        }
        catch (InvalidOperationException ex)
        {
            return NotFound(new { error = ex.Message });
        }
        catch (Exception)
        {
            return StatusCode(StatusCodes.Status500InternalServerError, new { error = "An unexpected error occurred" });
        }
    }

    [HttpGet]
    [ProducesResponseType(typeof(IEnumerable<UserResponse>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetAllUsers()
    {
        try
        {
            var users = await _unitOfWork.Users.GetAllAsync();
            var responses = _mapper.Map<IEnumerable<UserResponse>>(users);
            return Ok(responses);
        }
        catch (Exception)
        {
            return StatusCode(StatusCodes.Status500InternalServerError, new { error = "An unexpected error occurred" });
        }
    }
}

File: NotificationAPI.API/Controllers/UsersController.vb
Imports Microsoft.AspNetCore.Http
Imports Microsoft.AspNetCore.Mvc
Imports NotificationAPI.Application.DTOs.Requests
Imports NotificationAPI.Application.DTOs.Responses
Imports NotificationAPI.Application.UseCases.Users
Imports NotificationAPI.Domain.Entities
Imports NotificationAPI.Domain.Repositories
Imports AutoMapper

Namespace NotificationAPI.API.Controllers

    <ApiController>
    <Route("api/[controller]")>
    Public Class UsersController
        Inherits ControllerBase

        Private _unitOfWork As IUnitOfWork
        Private _mapper As IMapper
        Private _getUserByIdUseCase As GetUserByIdUseCase

        Public Sub New(unitOfWork As IUnitOfWork, mapper As IMapper, getUserByIdUseCase As GetUserByIdUseCase)
            _unitOfWork = unitOfWork
            _mapper = mapper
            _getUserByIdUseCase = getUserByIdUseCase
        End Sub

        <HttpPost>
        <ProducesResponseType(GetType(UserResponse), StatusCodes.Status201Created)>
        <ProducesResponseType(StatusCodes.Status400BadRequest)>
        Public Async Function CreateUser(<FromBody> request As CreateUserRequest) As Task(Of IActionResult)
            Try
                Dim user = _mapper.Map(Of User)(request)
                Await _unitOfWork.Users.AddAsync(user)
                Await _unitOfWork.SaveChangesAsync()

                Dim response = _mapper.Map(Of UserResponse)(user)
                Return CreatedAtAction(NameOf(GetUserById), New With {.id = response.Id}, response)
            Catch ex As Exception
                Return StatusCode(StatusCodes.Status500InternalServerError, New With {.error = "An unexpected error occurred"})
            End Try
        End Function

        <HttpGet("{id:guid}")>
        <ProducesResponseType(GetType(UserResponse), StatusCodes.Status200OK)>
        <ProducesResponseType(StatusCodes.Status404NotFound)>
        Public Async Function GetUserById(id As Guid) As Task(Of IActionResult)
            Try
                Dim response = Await _getUserByIdUseCase.ExecuteAsync(id)
                Return Ok(response)
            Catch ex As InvalidOperationException
                Return NotFound(New With {.error = ex.Message})
            Catch ex As Exception
                Return StatusCode(StatusCodes.Status500InternalServerError, New With {.error = "An unexpected error occurred"})
            End Try
        End Function

        <HttpGet>
        <ProducesResponseType(GetType(IEnumerable(Of UserResponse)), StatusCodes.Status200OK)>
        Public Async Function GetAllUsers() As Task(Of IActionResult)
            Try
                Dim users = Await _unitOfWork.Users.GetAllAsync()
                Dim responses = _mapper.Map(Of IEnumerable(Of UserResponse))(users)
                Return Ok(responses)
            Catch ex As Exception
                Return StatusCode(StatusCodes.Status500InternalServerError, New With {.error = "An unexpected error occurred"})
            End Try
        End Function
    End Class

End Namespace
